### ISSUE TITLE
[Bug]: editing device source produces error

### ISSUE THREAD

**Issue by jjmmfi** [josephdadams/TallyArbiter#217]
### What happened?

When I change settings on TallyArbiter I have gotten many error messages. One time the error came when I was trying to save settings I made and it did not save them. One time I guess it saved. I have just tested this app on 2 days. When I go to the report error page and press open a bug report, it gives me GitHub bad request: "Whoa there!
Your request URL is too long." It is frustrating that the "open error" pops up appears all the time I go to settings.

There is on error i got. Here are 2 more I got this time opening the app:
ReferenceError: sourceId is not defined
    at TallyArbiter_Edit_Device_Source (/Applications/TallyArbiter.app/Contents/Resources/app.asar/index.js:6597:50)
    at TallyArbiter_Manage (/Applications/TallyArbiter.app/Contents/Resources/app.asar/index.js:5619:14)
    at Socket.<anonymous> (/Applications/TallyArbiter.app/Contents/Resources/app.asar/index.js:1665:15)
    at Socket.emit (events.js:315:20)
    at Socket.emitUntyped (/Applications/TallyArbiter.app/Contents/Resources/app.asar/node_modules/socket.io/dist/typed-events.js:69:22)
    at /Applications/TallyArbiter.app/Contents/Resources/app.asar/node_modules/socket.io/dist/socket.js:428:39
    at processTicksAndRejections (internal/process/task_queues.js:75:11)

and:
TypeError: Cannot read property 'length' of undefined
    at UpdateDeviceState (/Applications/TallyArbiter.app/Contents/Resources/app.asar/index.js:5190:42)
    at InsertTestTally (/Applications/TallyArbiter.app/Contents/Resources/app.asar/index.js:1857:3)
    at TestTallies (/Applications/TallyArbiter.app/Contents/Resources/app.asar/index.js:1847:2)
    at EnableTestMode (/Applications/TallyArbiter.app/Contents/Resources/app.asar/index.js:1825:4)
    at Socket.<anonymous> (/Applications/TallyArbiter.app/Contents/Resources/app.asar/index.js:1712:4)
    at Socket.emit (events.js:315:20)
    at Socket.emitUntyped (/Applications/TallyArbiter.app/Contents/Resources/app.asar/node_modules/socket.io/dist/typed-events.js:69:22)
    at /Applications/TallyArbiter.app/Contents/Resources/app.asar/node_modules/socket.io/dist/socket.js:428:39
    at processTicksAndRejections (internal/process/task_queues.js:75:11)

Hope that helps. Running on Mac os Big Sur 11.2.3.

### Version

2.0.14 (2.0.14)

### Distribution

Desktop

### OS

MacOS

### What browsers are you seeing the problem on?

Chrome, Safari

### If applicable, What Listener Clients are You Using?

Web (Chrome)

### TallyArbiter configuration

```javascript
{
 "externalAddress": "http://0.0.0.0:4455/#/tally",
 "security": {
  "username_settings": "admin",
  "password_settings": "12345",
  "username_producer": "producer",
  "password_producer": "12345"
 },
 "sources": [
  {
   "data": {
    "me_onair": [
     "1"
    ],
    "cut_bus_mode": "off",
    "ip": "192.168.0.35"
   },
   "name": "ISO KAMERA",
   "sourceTypeId": "44b8bc4f",
   "reconnect": true,
   "enabled": true,
   "id": "4a6d19fd",
   "connected": true
  },
  {
   "data": {
    "ip": "192.168.0.35",
    "me_onair": [
     "1"
    ],
    "cut_bus_mode": "off"
   },
   "name": "PIENI KAMERA",
   "sourceTypeId": "44b8bc4f",
   "reconnect": true,
   "enabled": true,
   "id": "5b1bfe5c",
   "connected": true
  },
  {
   "data": {
    "ip": "192.168.0.35",
    "me_onair": [
     "1"
    ],
    "cut_bus_mode": "off"
   },
   "name": "KOLMAS KAMERA",
   "sourceTypeId": "44b8bc4f",
   "reconnect": true,
   "enabled": true,
   "id": "86bad5e7",
   "connected": true
  },
  {
   "data": {
    "password": "Voitto!",
    "ip": "192.168.0.30",
    "port": 4444
   },
   "name": "LIVE",
   "sourceTypeId": "4eb73542",
   "reconnect": true,
   "enabled": true,
   "id": "a9ca7fb2",
   "connected": false
  }
 ],
 "devices": [
  {
   "name": "ISO",
   "description": "",
   "tslAddress": 1,
   "enabled": true,
   "id": "62c2e2d4",
   "linkedPreview": true,
   "linkedProgram": true
  },
  {
   "name": "PIENI",
   "tslAddress": 2,
   "enabled": true,
   "id": "8797a486",
   "linkedProgram": true,
   "linkedPreview": false
  },
  {
   "name": "KOLMAS",
   "tslAddress": 3,
   "enabled": true,
   "id": "f7d717e0",
   "linkedProgram": true,
   "linkedPreview": false
  }
 ],
 "device_sources": [
  {
   "deviceId": "62c2e2d4",
   "sourceIdx": "0",
   "sourceId": "4a6d19fd",
   "id": "42503a60",
   "rename": true,
   "address": "6"
  },
  {
   "deviceId": "8797a486",
   "sourceIdx": "1",
   "sourceId": "5b1bfe5c",
   "id": "b8c86bd0",
   "rename": true,
   "address": "5"
  },
  {
   "deviceId": "f7d717e0",
   "sourceIdx": "2",
   "sourceId": "86bad5e7",
   "id": "0acf0c79",
   "address": "4",
   "rename": true
  },
  {
   "deviceId": "f7d717e0",
   "sourceIdx": "3",
   "address": "Blackmagic-laite",
   "rename": true,
   "sourceId": "a9ca7fb2",
   "id": "d0b6ca2e"
  },
  {
   "deviceId": "8797a486",
   "sourceIdx": "3",
   "address": "Blackmagic-laite",
   "sourceId": "a9ca7fb2",
   "id": "d572103e",
   "rename": true
  },
  {
   "deviceId": "62c2e2d4",
   "sourceIdx": "3",
   "address": "Blackmagic-laite",
   "sourceId": "a9ca7fb2",
   "id": "0b9bc5e8",
   "rename": true
  }
 ],
 "device_actions": [],
 "tsl_clients": [],
 "tsl_clients_1secupdate": false,
 "cloud_destinations": [],
 "cloud_keys": []
}
```

### Relevant log output

```shell
Logs:
[15.9.2021 klo 11.03.30] Loading the stored Tally Arbiter configuration file.
[15.9.2021 klo 11.03.30] The config file could not be found.
[15.9.2021 klo 11.03.30] Source Setup Complete.
[15.9.2021 klo 11.03.30] Initializing Device States.
[15.9.2021 klo 11.03.30] Device States Initialized.
[15.9.2021 klo 11.03.30] Setting up the REST API.
[15.9.2021 klo 11.03.30] REST API Setup Complete.
[15.9.2021 klo 11.03.30] Starting socket.IO Setup.
[15.9.2021 klo 11.03.30] Socket.IO Setup Complete.
[15.9.2021 klo 11.03.30] Starting OSC Setup.
[15.9.2021 klo 11.03.30] Starting VMix Emulation Service.
[15.9.2021 klo 11.03.30] Starting HTTP Server.
[15.9.2021 klo 11.03.30] OSC Sending Port Ready. Broadcasting on Port: 5958
[15.9.2021 klo 11.03.30] Finished VMix Emulation Setup. Listening for VMix Tally Connections on TCP Port 8099.
[15.9.2021 klo 11.03.30] Tally Arbiter running on port 4455
[15.9.2021 klo 11.14.55] Device Added: ISO KAMERA
[15.9.2021 klo 11.14.55] Config file saved to disk.
[15.9.2021 klo 11.15.19] Device Added: PIENI KAMERA
[15.9.2021 klo 11.15.19] Config file saved to disk.
[15.9.2021 klo 11.15.35] Device Added: KOLMAS KAMERA
[15.9.2021 klo 11.15.35] Config file saved to disk.
[15.9.2021 klo 11.15.46] Device Edited: PIENI KAMERA
[15.9.2021 klo 11.15.46] Config file saved to disk.
[15.9.2021 klo 11.15.54] Device Edited: ISO KAMERA
[15.9.2021 klo 11.15.54] Config file saved to disk.
[15.9.2021 klo 11.27.43] Config file saved to disk.
[15.9.2021 klo 11.27.49] Config file saved to disk.
[15.9.2021 klo 11.27.55] Config file saved to disk.
[15.9.2021 klo 11.28.06] Config file saved to disk.
[15.9.2021 klo 11.33.42] Source Added: ISO KAMERA
[15.9.2021 klo 11.33.42] Source: ISO KAMERA  Creating ATEM Connection.
[15.9.2021 klo 11.33.42] Config file saved to disk.
[15.9.2021 klo 11.34.44] Source Added: PIENI KAMERA
[15.9.2021 klo 11.34.44] Source: PIENI KAMERA  Creating ATEM Connection.
[15.9.2021 klo 11.34.44] Config file saved to disk.
[15.9.2021 klo 11.35.27] Source Added: KOLMAS KAMERA
[15.9.2021 klo 11.35.27] Source: KOLMAS KAMERA  Creating ATEM Connection.
[15.9.2021 klo 11.35.27] Config file saved to disk.
[15.9.2021 klo 11.37.36] Source Added: LIVE (KAMERAT)
[15.9.2021 klo 11.37.36] Source: LIVE (KAMERAT)  Creating OBS Websocket connection.
[15.9.2021 klo 11.37.36] Config file saved to disk.
[15.9.2021 klo 11.37.36] Source: LIVE (KAMERAT)  OBS Connection opened.
[15.9.2021 klo 11.37.36] OBS Tally Source: a9ca7fb2 Added new source: {{STREAMING}}
[15.9.2021 klo 11.37.36] OBS Tally Source: a9ca7fb2 Added new source: {{RECORDING}}
[15.9.2021 klo 11.37.36] Source: LIVE (KAMERAT)  OBS Authenticated.
[15.9.2021 klo 11.38.37] Device Source Added: ISO KAMERA - ISO KAMERA
[15.9.2021 klo 11.38.37] Config file saved to disk.
[15.9.2021 klo 11.38.49] Device Source Added: PIENI KAMERA - PIENI KAMERA
[15.9.2021 klo 11.38.49] Config file saved to disk.
[15.9.2021 klo 11.39.06] Device Source Added: KOLMAS KAMERA - KOLMAS KAMERA
[15.9.2021 klo 11.39.06] Config file saved to disk.
[15.9.2021 klo 11.39.46] Config file saved to disk.
[15.9.2021 klo 11.40.34] Config file saved to disk.
[15.9.2021 klo 11.41.52] Device Source Added: KOLMAS KAMERA - LIVE (KAMERAT)
[15.9.2021 klo 11.41.52] Config file saved to disk.
[15.9.2021 klo 11.45.05] Device Source Edited: KOLMAS KAMERA - KOLMAS KAMERA
[15.9.2021 klo 11.45.05] Config file saved to disk.
[15.9.2021 klo 11.45.31] Config file saved to disk.
[15.9.2021 klo 11.45.38] Config file saved to disk.
[15.9.2021 klo 11.46.10] Device Source Edited: KOLMAS KAMERA - LIVE (KAMERAT)
[15.9.2021 klo 11.46.10] Config file saved to disk.
[15.9.2021 klo 11.46.19] Source: LIVE (KAMERAT)  Program Scene Changed.
[15.9.2021 klo 11.46.19] OBS Tally Source: a9ca7fb2 Added new source: Spectralizer
[15.9.2021 klo 11.46.19] OBS Tally Source: a9ca7fb2 Added new source: Teksti
[15.9.2021 klo 11.46.19] OBS Tally Source: a9ca7fb2 Added new source: TIMER
[15.9.2021 klo 11.46.19] OBS Tally Source: a9ca7fb2 Added new source: 0
[15.9.2021 klo 11.46.19] OBS Tally Source: a9ca7fb2 Added new source: Untitled.mov
[15.9.2021 klo 11.46.19] OBS Tally Source: a9ca7fb2 Added new source: wipe.mov
[15.9.2021 klo 11.46.43] Source Edited: LIVE
[15.9.2021 klo 11.46.43] Config file saved to disk.
[15.9.2021 klo 12.51.45] Device Source Edited: KOLMAS KAMERA - LIVE
[15.9.2021 klo 12.51.45] Config file saved to disk.
[15.9.2021 klo 12.51.58] Device Source Edited: KOLMAS KAMERA - LIVE
[15.9.2021 klo 12.51.58] Config file saved to disk.
[15.9.2021 klo 12.52.04] Device Source Edited: KOLMAS KAMERA - LIVE
[15.9.2021 klo 12.52.04] Config file saved to disk.
[15.9.2021 klo 12.52.15] Device Source Edited: KOLMAS KAMERA - LIVE
[15.9.2021 klo 12.52.15] Config file saved to disk.
[15.9.2021 klo 12.52.20] Device Source Edited: KOLMAS KAMERA - LIVE
[15.9.2021 klo 12.52.20] Config file saved to disk.
[15.9.2021 klo 12.52.25] Device Edited: ISO KAMERA
[15.9.2021 klo 12.52.25] Config file saved to disk.
[15.9.2021 klo 12.52.35] Device Source Edited: KOLMAS KAMERA - LIVE
[15.9.2021 klo 12.52.35] Config file saved to disk.
[15.9.2021 klo 12.57.13] Caught exception: TypeError: Cannot read property 'length' of undefined
[15.9.2021 klo 12.58.21] Source: LIVE  Program Scene Changed.
[15.9.2021 klo 12.58.21] OBS Tally Source: a9ca7fb2 Added new source: Lower thirds
[15.9.2021 klo 12.58.21] OBS Tally Source: a9ca7fb2 Added new source: Blackmagic-laite
[15.9.2021 klo 12.59.11] Device Source Added: PIENI KAMERA - LIVE
[15.9.2021 klo 12.59.11] Config file saved to disk.
[15.9.2021 klo 12.59.32] Device Source Added: ISO KAMERA - LIVE
[15.9.2021 klo 12.59.32] Config file saved to disk.
[15.9.2021 klo 13.00.13] Device Source Edited: KOLMAS KAMERA - KOLMAS KAMERA
[15.9.2021 klo 13.00.13] Config file saved to disk.
[15.9.2021 klo 13.00.29] Device Source Edited: KOLMAS KAMERA - LIVE
[15.9.2021 klo 13.00.29] Config file saved to disk.
[15.9.2021 klo 13.00.59] Device Source Edited: PIENI KAMERA - PIENI KAMERA
[15.9.2021 klo 13.00.59] Config file saved to disk.
[15.9.2021 klo 13.01.09] Caught exception: ReferenceError: sourceId is not defined
[15.9.2021 klo 13.01.18] Caught exception: ReferenceError: sourceId is not defined
```

### Error stacktrace (if applicable)

```shell
ReferenceError: sourceId is not defined
    at TallyArbiter_Edit_Device_Source (/Applications/TallyArbiter.app/Contents/Resources/app.asar/index.js:6597:50)
    at TallyArbiter_Manage (/Applications/TallyArbiter.app/Contents/Resources/app.asar/index.js:5619:14)
    at Socket.<anonymous> (/Applications/TallyArbiter.app/Contents/Resources/app.asar/index.js:1665:15)
    at Socket.emit (events.js:315:20)
    at Socket.emitUntyped (/Applications/TallyArbiter.app/Contents/Resources/app.asar/node_modules/socket.io/dist/typed-events.js:69:22)
    at /Applications/TallyArbiter.app/Contents/Resources/app.asar/node_modules/socket.io/dist/socket.js:428:39
    at processTicksAndRejections (internal/process/task_queues.js:75:11)
```

**Comment by jeremiasmanninen** [comment]
[TALLY_ERRORS.zip](https://github.com/josephdadams/TallyArbiter/files/7195726/TALLY_ERRORS.zip)
Here is the whole TallyArbiter folder on a zip file, that contains all the error logs I got.

**Comment by hrueger** [comment]
Thanks for the error report. The not working bug report has been tracked in #187 , however, we haven't created a new release since then.

> It is frustrating that the "open error" pops up appears all the time I go to settings.

I've stumbled accross that problem, too, especially when developing. @MatteoGheza do you have time to fix that? I think we should
 (a) not show this warning when development mode is detected (`NODE_ENV` environment variable set to `development` and
 (b) only show a box in the settings window instead of a modal which needs to be dismissed.

Now to your last problem: The error is [here](https://github.com/josephdadams/TallyArbiter/blob/a318a0995e444fbcc09349a8180bbe0a47201d3a/index.js#L6672) for anyone wanting to make a PR. However, this is probably already fixed in #204 . That, however, needs some more testing before we can release it.

**Comment by josephdadams** [comment]
I would venture to say it's already fixed in the new upcoming branch or will be once we refactor that section of device-source assignments/mappings due to the new bus options coming. A workaround until that major refactor is released would be to just delete the device-source and recreate it.

**Comment by MatteoGheza** [comment]
I have a new idea for implementing bug reports (sending data compressed with LZMA-JS to the browser and then using GitHub Actions for de-compressing them when a new issue appears to have compressed data).
I already have a working PoC, but this week I don't have time to improve this.
I'll work on this ASAP.
Regarding the error popup, I'll work to improvements requested by @hrueger ASAP.

**Comment by Jocke4f** [comment]
Is this still a valid open issue? The above referenced issues https://github.com/josephdadams/TallyArbiter/issues/187 and https://github.com/josephdadams/TallyArbiter/pull/204 are assumed to have been fixed in 3.0.0.
